azkaban-common/src/main/java/azkaban/executor/ExecutorHealthChecker.java

```diff
@@ -169,11 +179,32 @@ private void handleExecutorNotAliveCase(
         (executor.getId(), 0) + 1);
     if (this.executorFailureCount.get(executor.getId()) % this.executorMaxFailureCount == 0
         && !this.alertEmails.isEmpty()) {
-      entry.getValue().stream().forEach(flow -> flow
-          .getExecutionOptions().setFailureEmails(this.alertEmails));
       logger.info(String.format("Executor failure count is %d. Sending alert emails to %s.",
           this.executorFailureCount.get(executor.getId()), this.alertEmails));
-      this.alerterHolder.get("email").alertOnFailedUpdate(executor, entry.getValue(), e);
+      this.alerterHolder.get("email").alertOnFailedExecutorHealthCheck(executor, entry.getValue(), e,
+          this.alertEmails);
+      this.cleanupForMissingExecutor(executor, entry.getValue());
+    }
+  }
+
+  /**
+   * Perform any cleanup required for an unreachable executor.
+   * Currently this accomplishes:
+   *   (1) Finalize the flow executions with appropriate reason
+   *   (2) Remove the corresponding executor entries from the data store.
+   *
+   * @param executor
+   * @param executions
+   */
+  private void cleanupForMissingExecutor(Executor executor, List<ExecutableFlow> executions) {
+    String finalizeReason = "Executor was unreachable, executor id: " + executor.getId();
+    finalizeFlows(executions, finalizeReason);
+
+    try {
+      this.executorLoader.removeExecutor(executor.getHost(), executor.getPort());
```

Comment:
Can you add a log line for removing executor adding the reason? I looked up removeExecutor and it does not log it.

Reply:
Will do.

Reply:
The behavior has been modified based on feedback and this is not needed anymore.